feat(ktreeelist): add component [khcp-5382] (#911)
Kong/kongponents: docs/components/tree-list.md

# Tree List

**KTreeList** - A drag-n-drop reorderable list component.

::: tip

> It can be tricky to determine if an item will be places below an item as a sibling or as a child when dropping. Items to be dropped as children should be dropped near the bottom border line of the parent. If this will be the first child of an item, you will notice the margin between the two items disappearing if you are hovering in the right place.

Reviewer: For this problem, I think increasing the drop zone height could give more space to get the feeling that the element will be placed there?

Author: If I update that it adds more margin between the items making it stop matching the design.

Author: 10px might be a big much, we have 4px right now. Here is how it would look with 10px.
![image](https://user-images.githubusercontent.com/67973710/203652401-32523856-625e-433c-8bbf-eca878032eeb.png)

> It can be tricky to determine if an item will be places below an item as a sibling or as a child when dropping. Items to be dropped as children should be dropped near the bottom border line of the parent. If this will be the first child of an item, you will notice the margin between the two items disappearing if you are hovering in the right place.

Reviewer 2: nit - replace `an item will be places` with `an item will be placed`

:::

<KTreeList :items="defaultItems" />

```html
<KTreeList :items="items" />
```

## Props

### items

An array of items that make up the tree.

<KTreeList :items="defaultItems2" />

```html
<template>
  <KTreeList :items="items" />
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  data() {
    return {
      items: [{
        name: "Cats",
        id: 'cats'
      },
      {
        name: "Dogs",
        id: 'dogs',
        children: [{
          name: "Puppies",
          id: 'puppies'
        }]
      },
      {
        name: "Bunnies",
        id: 'bunnies'
      }]
    }
  }
})
</script>
```

### disableDrag

Boolean (defaults to `false`) to turn of drag-n-drop reordering of the list.

<KTreeList disable-drag :items="disableItems" />

```html
<KTreeList disable-drag :items="items" />
```

### maxLevel

Use this prop to customize the maximum supported depth of the tree. We default to a max level of `2` or support for parents, children, and grandchildren.

<KTreeList :items="maxLevelItems" :max-level="1" />

```html
<KTreeList :items="items" :max-level="1" />
```

> ### v-model

Reviewer 3: If `v-model` usage is preferred, I think it should be listed above the `items` prop

`KTreeList` works with v-model for data binding.

::: danger
You cannot use `v-model` with the `items` prop. You must use one or the other.
:::

<div>
  <KLabel>Value:</KLabel> {{ myList }}
  <KTreeList v-model="myList" />
  <br>
  <KButton @click="reset">Reset</KButton>
</div>

```html
<template>
  <KLabel>Value:</KLabel> {{ myList }}
  <KTreeList v-model="myList" />
  <KButton @click="reset">Reset</KButton>
</template>

<script lang="ts">
import { defineComponent } from 'vue'

export default defineComponent({
  data() {
    return {
      myList: [{
        name: "Cats",
        id: 'cats'
      },
      {
        name: "Dogs",
        id: 'dogs',
        children: [{
          name: "Puppies",
          id: 'puppies'
        }]
      },
      {
        name: "Bunnies",
        id: 'bunnies'
      }]
    }
  },
  methods: {
    reset() {
      this.myList = [{
        name: "Cats",
        id: 'cats'
      },
      {
        name: "Dogs",
        id: 'dogs',
        children: [{
          name: "Puppies",
          id: 'puppies'
        }]
      },
      {
        name: "Bunnies",
        id: 'bunnies'
      }]
    }
  }
})
</script>
```

## KTreeItem

### Properties

> - `name` (required) - text displayed as the label for the item

Reviewer 3: I think these properties would go in the `items` prop documentation.

The `KTreeItem` component shouldn't be documented since its internal and cannot be used by a developer individually.

> - `id` (required) - a unique value used to identify the item

Reviewer 3: ```suggestion
- `id` (required) - a unique `string` used to identify the item
```

> - `selected` - boolean to indicate whether the current item is selected or not

Reviewer 3: I don't understand `selected` -- how is an item "selected" in the list..?

- `icon` - `KIcon` name to be displayed to the left of the item `name` (defaults to `treeDoc`, specify `none` to not display any icon)

> - `children` - an array of `KTreeItem`'s that will be styled as children of the current item

Reviewer 3: When you move this under the `items` prop, this would be changed to `an array of items that will be styled...`

### Item Slots

- `item-icon` - slot for content displayed to the left of the item name
- `item-label` - slot for the main content of an item (default's to display the `name` of the item)

> See the [Slots section](#Slots) for an example.

Reviewer 3: I don't think these would be documented anymore.

## Slots

`KTreeList` allows you to customize individual tree items via the item slots. These slots return the current `item` data as a slot param.

- `item-icon` - slot for content displayed to the left of the item name
- `item-label` - slot for the main content of an item (default's to display the `name` of the item)

<KTreeList :items="slotItems">
  <template #item-icon="{ item }">
    {{ item.id === 'cats' ? '😸' : item.id === 'bunnies' ? '🐰' : '🐶' }}
  </template>
  <template #item-label="{ item }">
    <span class="color-purple-400">
    {{ item.name }}
    </span>
  </template>
</KTreeList>

```html
<KTreeList :items="items">
  <template #item-icon="{ item }">
    {{ item.id === 'cats' ? '😸' : item.id === 'bunnies' ? '🐰' : '🐶' }}
  </template>
  <template #item-label="{ item }">
    <span class="color-purple-400">
    {{ item.name }}
    </span>
  </template>
</KTreeList>
```

## Events

- `@change` - Emitted when there is a change to the root level items
  - returns `items` - an array of `KTreeItem`s
- `@child-change` - Emitted when an item is added or removed at the non-root level

>   - returns `parent` - id of the parent item; `children` - an array of `KTreeItem`s

Reviewer 3: question: what if the parent doesn't have an `id`?

Author: All items are required to have a `name` and `id`, so it will.

>   - **Note:** two separate `child-change` events will fire if an item is moved from one parent to another

Reviewer 3: This is confusing to me. How will the consuming app know which event is correct? Is there a way to emit another value with the payload like `add` or `remove`?

Author: They are both correct, you are removing the item from one parent and adding it to another. The consuming app would need to update both `parent` items with the `children` arrays we provide on the event in order for the list to still be accurate. I tried simplifying this down to `add` / `remove` events but because we are piggybacking off the `draggable` events we are very limited in our options. I couldn't get that to work so instead I just provide you with the `id` of the item that changed and the updated `children` array.

Reviewer 3: I understand what you're saying, but again this seems really confusing to me without a fully working example in the docs.

Author: Added an example

  }]
}
</script>
```

### items

An array of items that make up the tree.

Item properties:

- `name` (required) - text displayed as the label for the item
- `id` (required) - a unique `string` used to identify the item
- `selected` - boolean to indicate whether the current item is selected or not
- `icon` - `KIcon` name to be displayed to the left of the item `name` (defaults to `treeDoc`, specify `none` to not display any icon)
- `children` - an array of items that will be styled as children of the current item

::: danger
You cannot use `v-model` with the `items` prop. You must use one or the other.
:::

<KTreeList :items="defaultItems2" />

```html
<template>
  <KTreeList :items="items" />
</template>

<script lang="ts" setup>
import { ref } from 'vue'

const items = ref([{
  name: "Cats",
  id: 'cats'
},
{
  name: "Dogs",
  id: 'dogs',
  children: [{
    name: "Puppies",
    id: 'puppies'
  }]
},
{
  name: "Bunnies",
  id: 'bunnies'
}])
</script>
```

### disableDrag

Boolean (defaults to `false`) to turn of drag-n-drop reordering of the list.

<KTreeList disable-drag :items="disableItems" />

```html
<KTreeList disable-drag :items="items" />
```

### maxLevel

> Use this prop to customize the maximum supported depth of the tree. We default to a max level of `2` or support for parents, children, and grandchildren.

Reviewer 3: > We default to a max level of `2` or support for parents, children, and grandchildren.

This sentence doesn't make sense to me? Typo maybe?

Reviewer 3: I'm still confused... if `maxLevel = 2` I would think that means parent and child (two levels)

But you're saying that allows you to have parent, child, and grandchild? (three levels)

Author: Fixed as we discussed on Friday

<KTreeList :items="maxLevelItems" :max-level="1" />

```html
<KTreeList :items="items" :max-level="1" />
```

## Slots

> `KTreeList` allows you to customize individual tree items via the item slots. These slots return the current `item` data as a slot param.

Reviewer 3: ```suggestion
`KTreeList` allows you to customize individual tree items via the item slots. The slots provide the current `item` data as a slot param.
```

> - `item-icon` - slot for content displayed to the left of the item name

Reviewer 3: ```suggestion
- `item-icon` - slot for content displayed to the left of the item name in place of the default icon
```

> - `item-icon` - slot for content displayed to the left of the item name

Reviewer 3: super-nit: When you documented the events, the sentence after the "dash" starts with a capital letter. 

All instances

> - `item-label` - slot for the main content of an item (default's to display the `name` of the item)

Reviewer 3: ```suggestion
- `item-label` - slot for the main content of an item (defaults to the `name` of the item)
```

<KTreeList :items="slotItems">
  <template #item-icon="{ item }">
    {{ item.id === 'cats' ? '😸' : item.id === 'bunnies' ? '🐰' : '🐶' }}

>   </template>

Reviewer 3: You may need to adjust the line-height a little for these emojis

  <template #item-label="{ item }">

>     <span class="color-purple-400">
>     {{ item.name }}
>     </span>

Reviewer 3: suggestion: To show that you're slotting something here besides just a color change, what if you added a prefix to the name or something as well?
```suggestion
    <span class="color-purple-400">
    Animal: {{ item.name }}
    </span>
```

(Don't forget to update the example too)

  </template>
</KTreeList>

```html
<KTreeList :items="items">
  <template #item-icon="{ item }">
    {{ item.id === 'cats' ? '😸' : item.id === 'bunnies' ? '🐰' : '🐶' }}
  </template>
  <template #item-label="{ item }">
    <span class="color-purple-400">
    {{ item.name }}
    </span>
  </template>
</KTreeList>
```

## Events

- `@change` - Emitted when there is a change to the root level items
  - returns `items` - an array of tree items
- `@child-change` - Emitted when an item is added or removed at the non-root level
  - returns `parent` - id of the parent item; `children` - an array of tree items
  - **Note:** two separate `child-change` events will fire if an item is moved from one parent to another

> - `@selected` - Emitted when you click (and don't drag-n-drop) an item; returns the selected item's data

Reviewer 3: What if you click and then drag? Does it only fire on click and release?
